### PR TITLE
Gitlab backend

### PR DESCRIPTION
**Pull request by assadOW2**
I added a new backend, Gitlab, based on Github. Metadata and urls are different but overall works the same way as github.

### CODE REVIEW

File: perceval/backends/core/gitlab.py [hunk 1]
```diff
@@ -0,0 +1,689 @@
...
+
+    def __get_issue_assignees(self, raw_assignees):
+        """Get issue assignees"""
+
+        self._push_cache_queue('{ASSIGNEES}')
+        self._push_cache_queue(raw_assignees)
+        self._flush_cache_queue()
+        assignees = []        
+        for ra in raw_assignees:
+            if 'login' in ra.keys():
```
Comment (valeriocos) [inline perceval/backends/core/gitlab.py]: I don't see in the GitLab API the attribute 'login' for assignees. If this is the case, the first if condition can be removed

**Comment by assadOW2** [comment]
@valeriocos : about the attribute 'login' in assignees in ___get_issue_assignees_. 
You are absolutely right. At the beginning, I did updates on github.py, before making gitlab own script. So this is an attribute I forgot to remove.

**Comment by assadOW2** [comment]
@jgbarah : I will take a look on flake8 warning. About the tests, I did them on OW2 gitlab with my personal token. I wouldn't refuse your help in writing them.

**Comment by jgbarah** [comment]
@assadOW2 said:

> About the tests, I did them on OW2 gitlab with my personal token. I wouldn't refuse your help in writing them.

If you can, please have a look at the tests. Usually, to avoid connecting to the real thing every time we unit test, what we do is to retrieve data from the API once, and then using httpretty, access the downloaded data. The library "thinks" it is accessing the real API, but it is just using httpretty to access the "fake" data.

If you want, you can have a look at the tests for GitHub, or let me know, and I try to do the tests myself.

**Comment by assadOW2** [comment]
@jgbarah : I will take a look on the GitHub tests and if I have any trouble, I will let you know. Thanks.

**Comment by valeriocos** [comment]
@assadOW2 thank you for your contribution!
I had a look to the GitLab API and found some differences with the GitHub one. For instance:
- GitLab issues already include user data (instead in GitHub we need to make an explicit call to the API)

- Reactions, called **award_emoji** in GitLab, exist but need to be fetched in a different way. A GitHub issue comes with the number of reactions, thus we can make a API call only if this number is not zero. Conversely, GitLab issues don't seem to come with the number, so we have always to make a call to the API

- Responses of the GitLab API don't seem come with a RateLimit-Reset, so we have to set the sleeping time (**seconds_to_reset**) to a predefined value.

Due to these differences I thought that was better to send you a restyled version of your code. Feel free to use it and modified (there are still quite a lot of improvements to do :) ).

```python
# -*- coding: utf-8 -*-
#
# Copyright (C) 2015-2017 Bitergia
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
#
# Authors:
#     ...
#

import json
import logging
import time

import requests

from grimoirelab.toolkit.datetime import datetime_to_utc, str_to_datetime
from grimoirelab.toolkit.uris import urijoin

from ...backend import (Backend,
                        BackendCommand,
                        BackendCommandArgumentParser,
                        metadata)
from ...errors import CacheError, RateLimitError
from ...utils import DEFAULT_DATETIME

GITLAB_URL = "https://gitlab.com/"
GITLAB_API_URL = "https://gitlab.com/api/v4"

# Range before sleeping until rate limit reset
MIN_RATE_LIMIT = 10
MAX_RATE_LIMIT = 500

TARGET_ISSUE_FIELDS = ['user_notes_count', 'award_emoji']

logger = logging.getLogger(__name__)


class GitLab(Backend):
    """GitLab backend for Perceval.

    This class allows the fetch the issues stored in GitLab
    repository.

    :param owner: GitLab owner
    :param repository: GitLab repository from the owner
    :param api_token: GitLab auth token to access the API
    :param tag: label used to mark the data
    :param cache: use issues already retrieved in cache
    :param sleep_for_rate: sleep until rate limit is reset
    :param min_rate_to_sleep: minimun rate needed to sleep until
         it will be reset
    """
    version = '0.1.0'

    def __init__(self, owner=None, repository=None,
                 api_token=None,
                 tag=None, cache=None,
                 sleep_for_rate=False, min_rate_to_sleep=MIN_RATE_LIMIT):
        origin = urijoin(GITLAB_URL, owner, repository)

        super().__init__(origin, tag=tag, cache=cache)
        self.owner = owner
        self.repository = repository
        self.api_token = api_token
        self.client = GitLabClient(owner, repository, api_token,
                                   sleep_for_rate, min_rate_to_sleep)
        self._users = {}  # internal users cache

    @classmethod
    def has_caching(cls):
        """Returns whether it supports caching items on the fetch process.

        :returns: this backend supports items cache
        """
        return True

    @classmethod
    def has_resuming(cls):
        """Returns whether it supports to resume the fetch process.

        :returns: this backend supports items resuming
        """
        return True

    @staticmethod
    def metadata_id(item):
        """Extracts the identifier from a GitLab item."""

        return str(item['id'])

    @staticmethod
    def metadata_updated_on(item):
        """Extracts the update time from a GitLab item.

        The timestamp used is extracted from 'updated_at' field.
        This date is converted to UNIX timestamp format. As GitLab
        dates are in UTC the conversion is straightforward.

        :param item: item generated by the backend

        :returns: a UNIX timestamp
        """
        ts = item['updated_at']
        ts = str_to_datetime(ts)

        return ts.timestamp()

    @staticmethod
    def metadata_category(item):
        """Extracts the category from a GitLab item.

        This backend only generates one type of item which is
        'issue'.
        """
        return 'issue'

    @metadata
    def fetch(self, from_date=DEFAULT_DATETIME):
        """Fetch the issues from the repository.

        The method retrieves, from a GitLab repository, the issues
        updated since the given date.

        :param from_date: obtain issues updated since this date

        :returns: a generator of issues
        """

        self._purge_cache_queue()

        from_date = datetime_to_utc(from_date)

        issues_groups = self.client.issues(start=from_date)

        for raw_issues in issues_groups:
            self._push_cache_queue('{ISSUES}')
            self._push_cache_queue(raw_issues)
            self._flush_cache_queue()
            issues = json.loads(raw_issues)
            for issue in issues:
                self.__init_extra_issue_fields(issue)

                issue['user_notes_data'] = self.__get_issue_notes(issue['iid'])
                issue['award_emoji_data'] = self.__get_issue_award_emoji(issue['iid'])

                self._push_cache_queue('{ISSUE-END}')
                self._flush_cache_queue()

                yield issue
        self._push_cache_queue('{}{}')
        self._flush_cache_queue()

    @metadata
    def fetch_from_cache(self):
        """Fetch the issues from the cache.
        It returns the issues stored in the cache object provided during
        the initialization of the object. If this method is called but
        no cache object was provided, the method will raise a `CacheError`
        exception.
        :returns: a generator of items
        :raises CacheError: raised when an error occurs accessing the
            cache
        """
        if not self.cache:
            raise CacheError(cause="cache instance was not provided")

        cache_items = self.cache.retrieve()
        raw_item = next(cache_items)

        while raw_item != '{}{}':

            if raw_item == '{ISSUES}':
                issues = self.__fetch_issues_from_cache(cache_items)

            for issue in issues:
                self.__init_extra_issue_fields(issue)
                raw_item = next(cache_items)

                while raw_item != '{ISSUE-END}':
                    try:
                        if raw_item == '{NOTES}':
                            notes = self.__fetch_notes_from_cache(cache_items)
                            issue['user_notes_data'] = notes
                        elif raw_item == '{ISSUE-EMOJIS}':
                            emoji = self.__fetch_issue_emoji_from_cache(cache_items)
                            issue['award_emoji_data'] = emoji
                        raw_item = next(cache_items)

                    except StopIteration:
                        # this should be never executed,
                        # the while condition prevents
                        # to trigger the StopIteration exception
                        break

                raw_item = next(cache_items)
                yield issue

    def __get_issue_notes(self, issue_id):
        """Get issue notes"""

        notes = []

        group_notes = self.client.issue_notes(issue_id)
        self._push_cache_queue('{NOTES}')
        self._flush_cache_queue()

        for raw_notes in group_notes:
            self._push_cache_queue(raw_notes)
            self._flush_cache_queue()

            for note in json.loads(raw_notes):
                note_id = note['id']
                note['award_emoji_data'] = self.__get_note_award_emoji(issue_id, note_id)
                notes.append(note)

        return notes

    def __get_issue_award_emoji(self, issue_id):
        """Get award emojis for issue"""

        emojis = []
        self._push_cache_queue('{ISSUE-EMOJIS}')
        self._flush_cache_queue()

        group_emojis = self.client.issue_emojis(issue_id)
        for raw_emojis in group_emojis:
            self._push_cache_queue(raw_emojis)
            self._flush_cache_queue()

            for emoji in json.loads(raw_emojis):
                emojis.append(emoji)

        return emojis

    def __get_note_award_emoji(self, issue_id, note_id):
        """Fetch emojis for note"""

        emojis = []
        self._push_cache_queue('{NOTE-EMOJIS}')
        self._flush_cache_queue()

        group_emojis = self.client.note_emojis(issue_id, note_id)
        for raw_emojis in group_emojis:
            self._push_cache_queue(raw_emojis)
            self._flush_cache_queue()

            for emoji in json.loads(raw_emojis):
                emojis.append(emoji)

        return emojis

    def __fetch_issues_from_cache(self, cache_items):
        """Fetch issues from cache"""

        raw_content = next(cache_items)
        issues = json.loads(raw_content)
        return issues

    def __fetch_issue_emoji_from_cache(self, cache_items):
        """Fetch issue emoji from cache"""

        raw_content = next(cache_items)
        emoji = json.loads(raw_content)
        return emoji

    def __fetch_notes_from_cache(self, cache_items):
        """Fetch issue notes from cache"""

        raw_content = next(cache_items)
        notes = json.loads(raw_content)

        for note in notes:
            _ = next(cache_items)
            raw_emoji = next(cache_items)
            note['award_emoji_data'] = json.loads(raw_emoji)

        return notes

    def __init_extra_issue_fields(self, issue):
        """Add fields to an issue"""

        issue['user_notes_data'] = []
        issue['award_emoji_data'] = []


class GitLabClient:
    """Client for retieving information from GitLab API"""

    MAX_RETRIES = 5
    _users = {}       # users cache
    _users_orgs = {}  # users orgs cache

    def __init__(self, owner, repository, token,
                 sleep_for_rate=False, min_rate_to_sleep=MIN_RATE_LIMIT):
        self.owner = owner
        self.repository = repository
        self.token = token
        self.rate_limit = None
        self.sleep_for_rate = sleep_for_rate

        self.api_url = GITLAB_API_URL

        if min_rate_to_sleep > MAX_RATE_LIMIT:
            msg = "Minimum rate to sleep value exceeded (%d)."
            msg += "High values might cause the client to sleep forever."
            msg += "Reset to %d."
            logger.warning(msg, min_rate_to_sleep, MAX_RATE_LIMIT)

        self.min_rate_to_sleep = min(min_rate_to_sleep, MAX_RATE_LIMIT)

    def issue_notes(self, issue_id):
        """Get the issue notes from pagination"""

        return self._fetch_items(urijoin("issues", str(issue_id), "notes"), "note")

    def issues(self, start=None):
        """Get the issues from pagination"""

        return self._fetch_items("issues", payload_type="issue", start=start)

    def issue_emojis(self, issue_id):
        """Get emojis of an issue"""

        return self._fetch_items(urijoin("issues", str(issue_id), "award_emoji"), "emoji")

    def note_emojis(self, issue_id, note_id):
        """Get emojis of a note"""

        return self._fetch_items(urijoin("issues", str(issue_id), "notes", str(note_id), "award_emoji"), "emoji")

    def user(self, username):
        """Get the user information and update the user cache"""

        if username in self._users:
            return self._users[username]

        url_user = urijoin(self.api_url, 'users?username=' + username)

        logging.info("Getting info for %s" % (url_user))
        r = self.__send_request(url_user, headers=self.__build_headers())
        user = r.text
        self._users[username] = user

        return user

    def __get_url_repo(self):
        """Build URL repo"""

        url_repo = \
            urijoin(
                self.api_url, 'projects', self.owner + '%2F' + self.repository)

        return url_repo

    def __get_url(self, path, startdate=None):
        """Build repo-"related URL"""

        url = urijoin(self.__get_url_repo(), path)
        return url

    def __build_payload(self, payload_type='issue', startdate=None):
        """Build payload"""

        # Gitlab sort arguments
        payload = {'sort': 'asc'}
        if payload_type == 'issue':
            payload['state'] = 'all'
            payload['order_by'] = 'updated_at'
        elif payload_type in ['note', 'emoji']:
            payload['order_by'] = 'updated_at'

        if startdate:
            startdate = startdate.isoformat()
            payload['created_after'] = startdate

        return payload

    def __build_headers(self):
        """Set header for request"""
        if self.token:
            headers = {'PRIVATE-TOKEN': self.token}

            return headers

    def __send_request(self, url, params=None, headers=None):
        """GET HTTP caring of rate limit"""

        retries = 0

        while retries < self.MAX_RETRIES:
            try:
                if self.rate_limit is not None \
                        and self.rate_limit <= self.min_rate_to_sleep:

                    seconds_to_reset = 3600
                    cause = "GitLab rate limit exhausted."
                    if self.sleep_for_rate:
                        logger.info("%s Waiting %i secs for rate limit reset.",
                                    cause, seconds_to_reset)
                        time.sleep(seconds_to_reset)
                    else:
                        raise RateLimitError(
                            cause=cause, seconds_to_reset=seconds_to_reset)

                r = requests.get(url, params=params, headers=headers)
                r.raise_for_status()

                # GitLab service establishes API rate limits.
                if 'RateLimit-Remaining' in r.headers:
                    self.rate_limit = int(r.headers['RateLimit-Remaining'])
                    logger.debug("Rate limit: %s", self.rate_limit)
                else:
                    self.rate_limit = None

                break
            except requests.exceptions.HTTPError as e:
                if e.response.status_code == 403:
                    retry_seconds = int(10)
                    logger.warning(
                        "Abuse rate limit, the backend will sleep for %s seconds \
                            before starting again",
                        retry_seconds)
                    time.sleep(retry_seconds)
                    retries += 1
                else:
                    raise e

        if retries == self.MAX_RETRIES:
            r.raise_for_status()

        return r

    def _fetch_items(self, path, payload_type, start=None):
        """Return the items from gitalb API using links pagination"""

        page = 0  # current page
        last_page = None  # last page
        url_next = self.__get_url(path, start)

        logger.debug("Get GitLab paginated items from " + url_next)
        payload = self.__build_payload(payload_type, start)
        r = self.__send_request(url_next, payload, self.__build_headers())
        items = r.text
        page += 1

        if 'last' in r.links:
            last_url = r.links['last']['url']
            last_page = last_url.split('&page=')[1].split('&')[0]
            last_page = int(last_page)
            logger.debug("Page: %i/%i" % (page, last_page))

        while items:
            yield items

            items = None

            if 'next' in r.links:
                url_next = r.links['next']['url']  # Loving requests :)
                r = self.__send_request(
                    url_next, payload, self.__build_headers())
                page += 1
                items = r.text
                logger.debug("Page: %i/%i" % (page, last_page))


class GitLabCommand(BackendCommand):
    """Class to run GitLab backend from the command line."""

    BACKEND = GitLab

    @staticmethod
    def setup_cmd_parser():
        """Returns the GitLab argument parser."""

        parser = BackendCommandArgumentParser(from_date=True,
                                              token_auth=True,
                                              cache=True)

        # GitLab options
        group = parser.parser.add_argument_group('GitLab arguments')
        group.add_argument('--sleep-for-rate', dest='sleep_for_rate',
                           action='store_true',
                           help="sleep for getting more rate")
        group.add_argument('--min-rate-to-sleep', dest='min_rate_to_sleep',
                           default=MIN_RATE_LIMIT, type=int,
                           help="sleep until reset when the rate limit \
                               reaches this value")

        # Positional arguments
        parser.parser.add_argument('owner',
                                   help="GitLab owner")
        parser.parser.add_argument('repository',
                                   help="GitLab repository")

        return parser

```

**Comment by valeriocos** [comment]
If you want to dive in testing the backend, I have prepared the httpretty server and sketched some tests to do. The code below has to be added to [tests/test_gitlab.py](https://github.com/grimoirelab/perceval/tree/master/tests), while the zip ([available in google drive](https://drive.google.com/open?id=13gYADfu5zP2QryYOMSWP1hRgxTEF_iJQ)) should be put in the directory [tests/data](https://github.com/grimoirelab/perceval/tree/master/tests/data). 

Let me know if you have some questions about the backend or the tests :)

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright (C) 2015-2017 Bitergia
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
#
# Authors:
#     ...
#

import datetime
import httpretty
import json
import os
import pkg_resources
import shutil
import sys
import tempfile
import time
import unittest

import requests


# Hack to make sure that tests import the right packages
# due to setuptools behaviour
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
pkg_resources.declare_namespace('perceval.backends')

from perceval.backend import BackendCommandArgumentParser
from perceval.cache import Cache
from perceval.errors import CacheError, RateLimitError
from perceval.utils import DEFAULT_DATETIME
from perceval.backends.core.gitlab import (GitLab,
                                           GitLabCommand,
                                           GitLabClient)


GITLAB_API_URL = "https://gitlab.com/api/v4"
GITLAB_ISSUES_URL = GITLAB_API_URL + "/projects/fdroid%2Ffdroiddata/issues"

def read_file(filename, mode='r'):
    with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), filename), mode) as f:
        content = f.read()
    return content


def setup_http_server():

    page_1 = read_file('data/gitlab/issue_page_1')
    page_2 = read_file('data/gitlab/issue_page_2')

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL,
                           body=page_1,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20',
                               'Link': '<' + GITLAB_ISSUES_URL + '/?&page=2>; rel="next", <' +
                                       GITLAB_ISSUES_URL + '/?&page=3>; rel="last"'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + '/?&page=2',
                           body=page_2,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    issue_1_notes = read_file('data/gitlab/issue_1_notes')
    issue_2_notes = read_file('data/gitlab/issue_2_notes')
    issue_3_notes = read_file('data/gitlab/issue_3_notes')
    issue_4_notes = read_file('data/gitlab/issue_4_notes')

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/1/notes",
                           body=issue_1_notes,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/2/notes",
                           body=issue_2_notes,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/3/notes",
                           body=issue_3_notes,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/4/notes",
                           body=issue_4_notes,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    emoji = read_file('data/gitlab/emoji')
    empty_emoji = read_file('data/gitlab/empty_emoji')

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/1/award_emoji",
                           body=emoji,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/2/award_emoji",
                           body=empty_emoji,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/3/award_emoji",
                           body=emoji,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/4/award_emoji",
                           body=empty_emoji,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/1/notes/1/award_emoji",
                           body=emoji,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/1/notes/2/award_emoji",
                           body=empty_emoji,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/2/notes/1/award_emoji",
                           body=empty_emoji,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/3/notes/1/award_emoji",
                           body=empty_emoji,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

    httpretty.register_uri(httpretty.GET,
                           GITLAB_ISSUES_URL + "/4/notes/1/award_emoji",
                           body=empty_emoji,
                           status=200,
                           forcing_headers={
                               'RateLimit-Remaining': '20'
                           })

class TestGitHubBackend(unittest.TestCase):
    """ GitHub backend tests """

    def test_initialization(self):
        """Test whether attributes are initializated"""

        gitlab = GitLab('fdroid', 'fdroiddata', 'aaa', tag='test')

        self.assertEqual(gitlab.owner, 'fdroid')
        self.assertEqual(gitlab.repository, 'fdroiddata')
        self.assertEqual(gitlab.origin, 'https://gitlab.com/fdroid/fdroiddata')
        self.assertEqual(gitlab.tag, 'test')

        # When tag is empty or None it will be set to
        # the value in origin
        gitlab = GitLab('fdroid', 'fdroiddata', 'aaa')
        self.assertEqual(gitlab.owner, 'fdroid')
        self.assertEqual(gitlab.repository, 'fdroiddata')
        self.assertEqual(gitlab.origin, 'https://gitlab.com/fdroid/fdroiddata')
        self.assertEqual(gitlab.tag, 'https://gitlab.com/fdroid/fdroiddata')

        gitlab = GitLab('fdroid', 'fdroiddata', 'aaa', tag='')
        self.assertEqual(gitlab.owner, 'fdroid')
        self.assertEqual(gitlab.repository, 'fdroiddata')
        self.assertEqual(gitlab.origin, 'https://gitlab.com/fdroid/fdroiddata')
        self.assertEqual(gitlab.tag, 'https://gitlab.com/fdroid/fdroiddata')

    def test_has_caching(self):
        """Test if it returns True when has_caching is called"""

        self.assertEqual(GitLab.has_caching(), True)

    def test_has_resuming(self):
        """Test if it returns True when has_resuming is called"""

        self.assertEqual(GitLab.has_resuming(), True)

    @httpretty.activate
    def test_fetch(self):
        """ Test whether a list of issues is returned """

        setup_http_server()

        backend = GitLab("fdroid", "fdroiddata", "your-token")
        issues = [issues for issues in backend.fetch()]

        self.assertEqual(len(issues), 4)

    @httpretty.activate
    def test_fetch_from_date(self):
        """ Test when return from date """

        print("TODO")

    @httpretty.activate
    def test_fetch_empty(self):
        """ Test when return empty """

        print("TODO")

    # def test_live(self):
    #     from_date = datetime.datetime(2017, 4, 10)
    #     backend = GitLab("fdroid", "fdroiddata", "your-token")
    #     issues = [issues for issues in backend.fetch(from_date=from_date)]


class TestGitHubBackendCache(unittest.TestCase):
    """GitHub backend tests using a cache"""

    def setUp(self):
        self.tmp_path = tempfile.mkdtemp(prefix='perceval_')

    def tearDown(self):
        shutil.rmtree(self.tmp_path)

    @httpretty.activate
    def test_fetch_from_cache(self):
        """ Test whether a list of issues is returned from cache """

        print("TODO")

    def test_fetch_from_empty_cache(self):
        """Test if there are not any issues returned when the cache is empty"""

        print("TODO")

    def test_fetch_from_non_set_cache(self):
        """Test if a error is raised when the cache was not set"""

        print("TODO")


class TestGitHubClient(unittest.TestCase):
    """ GitHub API client tests """

    def test_api_url_initialization(self):
        """Test API URL initialization for both basic and enterprise servers"""

        print("TODO")

    @httpretty.activate
    def test_issues(self):
        """ Test issues API call """

        print("TODO")

    @httpretty.activate
    def test_issues_from_date(self):
        """ Test issue from date param API call """

        print("TODO")

    @httpretty.activate
    def test_issues_empty(self):
        """ Test when issue is empty API call """

        print("TODO")

    @httpretty.activate
    def test_issue_notes(self):
        """ Test issue_notes API call """

        print("TODO")

    @httpretty.activate
    def test_issue_emoji(self):
        """ Test issue_emoji API call """

        print("TODO")

    @httpretty.activate
    def test_note_emojis(self):
        """ Test note_emoji API call """

        print("TODO")

    @httpretty.activate
    def test_user(self):
        """Test user API call"""

        print("TODO")

    @httpretty.activate
    def test_http_wrong_status(self):
        """Test if a error is raised when the http status was not 200"""

        print("TODO")

    @httpretty.activate
    def test_sleep_for_rate(self):
        """ Test get_page_issue API call """

        print("TODO")

    @httpretty.activate
    def test_rate_limit_error(self):
        """ Test get_page_issue API call """

        print("TODO")


class TestGitLabCommand(unittest.TestCase):
    """GitHubCommand unit tests"""

    def test_backend_class(self):
        """Test if the backend class is GitHub"""

        self.assertIs(GitLabCommand.BACKEND, GitLab)

    def test_setup_cmd_parser(self):
        """Test if it parser object is correctly initialized"""

        parser = GitLabCommand.setup_cmd_parser()
        self.assertIsInstance(parser, BackendCommandArgumentParser)

        args = ['--sleep-for-rate',
                '--min-rate-to-sleep', '1',
                '--tag', 'test', '--no-cache',
                '--api-token', 'abcdefgh',
                '--from-date', '1970-01-01',
                'fdroid', 'fdroiddata']

        parsed_args = parser.parse(*args)
        self.assertEqual(parsed_args.owner, 'fdroid')
        self.assertEqual(parsed_args.repository, 'fdroiddata')
        self.assertEqual(parsed_args.sleep_for_rate, True)
        self.assertEqual(parsed_args.min_rate_to_sleep, 1)
        self.assertEqual(parsed_args.tag, 'test')
        self.assertEqual(parsed_args.from_date, DEFAULT_DATETIME)
        self.assertEqual(parsed_args.no_cache, True)
        self.assertEqual(parsed_args.api_token, 'abcdefgh')


if __name__ == "__main__":
    unittest.main(warnings='ignore')

``

**Comment by assadOW2** [comment]
@valeriocos : thanks for the work ! I noticed you removed entreprise-url argument. In our case (at OW2) we use an enterprise url so I think it's better to have this option. I will add to your code this argument if you don't mind.
Another point; I was struggling these last days on the tests, having hard time making httpretty work. Is it possible to contact one of you about this ? Thanks in advance

**Comment by valeriocos** [comment]
@assadOW2 you're welcome, feel free to modify the code and integrate it in your PR (sorry for removing the enterprise-url argument).
You can write me at valcos@bitergia.com

**Comment by jgbarah** [comment]
@sduenas @valeriocos, any feedback about this?

**Comment by valeriocos** [comment]
There is another [PR](https://github.com/grimoirelab/perceval/pull/252), which completes the tests and extends the code with the generic HTTP client.

**Pull request closed without merging**